3d driving

ethan10mak opened the pull request:
### Description
This PR adds support for 3D mode.
- Elevation parsed from Open Drive files
- Added mesh generation through the Trimesh library
- Tested driving simulations through CARLA
- Updated older test cases to generate in 2D mode

### Issue Link
N/A

### Checklist
- [x] I have tested the changes locally via `pytest` and/or other means
- [x] I have added or updated relevant documentation
- [x] I have autoformatted the code with black and isort
- [x] I have added test cases (if applicable)

### Additional Notes
N/A

ethan10mak committed:
Update xodr_parser.py

Made comments to file

ethan10mak committed:
Added parse_elevation_profile code

ethan10mak committed:
elevationProfile and lateralProfile parsing implementation

Started writing code to parse elevationProfile and lateralProfile elements.

ethan10mak committed:
Changed cur_p to have three dimensions

ethan10mak committed:
Formatted xodr_parser.py

ethan10mak committed:
Revert "Formatted xodr_parser.py"

This reverts commit 72a31f2d79b794f3374d09143477dddcaf0419c8.

ethan10mak committed:
Updated formatting of xodr_parser.py

ethan10mak committed:
Added get_super_elevation_at function

ethan10mak committed:
Update xodr_parser.py

ethan10mak committed:
Updated parser, regions, and geometry to account for elevation

ethan10mak committed:
Updating xodr_parser

Trying to take 3 dimensional vectors and display them, currently displays visual window, but doesn't show anything

ethan10mak committed:
Modifying xodr_parser to accept 3D and create trimeshes

Doesn't quite work yet, but I have differentiated functions such as calc_geometry_for_type() to calculate different things for 2D and 3D

ethan10mak committed:
Updating calculate_geometry and toScenicRoad to 3D

Doesn't work yet, currently facing an error where PolylineRegion and PathRegion are not compatible.

ethan10mak committed:
Visualized working meshes

ethan10mak committed:
Updated NetworkElement to Region

ethan10mak committed:
3D meshes outputs properly

ethan10mak committed:
Updated roads.py

ethan10mak committed:
Updated findPointIn Function

ethan10mak committed:
updated global param use2DMap

ethan10mak committed:
Updating use2DMap param

ethan10mak committed:
Updated abstract function and orientation

ethan10mak committed:
Updated abstract classes

ethan10mak committed:
Driving Orientation Fixed and 2D Compatibility

ethan10mak committed:
Updated 2D Compatibility

ethan10mak committed:
Updated documentation

ethan10mak committed:
Added __getitem__ to PathRegion

ethan10mak committed:
Updated behaviors.py for 3d concatenateCenterlines

ethan10mak committed:
Updated xodr_parser.py

Made sure that all PathRegions created are created through points

ethan10mak committed:
Updated behavior.scenic

Updated the FollowLaneBehavior function to deal with 2d and 3d points

ethan10mak committed:
Added signedDistanceTo() to PathRegion

ethan10mak committed:
Updated curbRegion

ethan10mak committed:
Updated for optimized region usage.

ethan10mak committed:
Intersection are generated correctly

ethan10mak committed:
Fixed problems with orientation

Orientation with "on Road" should function as intended

ethan10mak committed:
Updated Test

ethan10mak committed:
Squashed commit of the following:

commit e0d70930d1321e2033828e33cb11e2c0ab8908b8
Author: Lola Marrero <110120745+lola831@users.noreply.github.com>
Date:   Tue Jun 17 21:06:38 2025 +0000

    Remove unused CameraManager.images to fix memory leak (#353)

commit 14cba62590f12c54ac6ea9073eaf703123cb00af
Author: Eric Vin <8935814+Eric-Vin@users.noreply.github.com>
Date:   Tue Jun 10 14:13:46 2025 -0400

    PathRegion/PolygonalFootprint Test and Bugfix (#351)

    * Added test/bugfix for intersection of PathRegion and PolygonalFootprint

    * Formatting trigger

    * Manual formatting.

commit 5ebdf46c49a548fe00dba374b0d5d9078fb80464
Author: Lola Marrero <110120745+lola831@users.noreply.github.com>
Date:   Mon Jun 2 22:28:03 2025 +0000

    Run tests weekly + Slack notification (#347)

    * weekly ci tests

    * Finish weekly ci tests workflow

commit 44a844279f840498053b2909fe8d5bfa700d54cb
Author: Lola Marrero <110120745+lola831@users.noreply.github.com>
Date:   Fri May 23 16:52:04 2025 +0000

    Constrain pyglet to >=1.5, <=1.5.26 to avoid trimesh and macOS compatibility issues (#346)

    * Pin pyglet to ~=1.5 to avoid incompatibility with trimesh

    * changed to pyglet >= 1.5, <= 1.5.26

commit c1003e9b89c612e81790151640fb78f298a7782d
Author: Lola Marrero <110120745+lola831@users.noreply.github.com>
Date:   Mon May 12 13:52:56 2025 -0700

    Add dynamic pedestrian example and helper function; render MetaDrive 3D without added terrain or physics (#343)

commit 17404e0ae2deea0b7b144104a44213eca1173033
Author: Lola Marrero <110120745+lola831@users.noreply.github.com>
Date:   Tue Apr 29 15:53:39 2025 -0700

    Add support for initial velocity in MetaDrive simulations (#340)

commit 56608bacbc8f59ceab0420da439669e69b5221cd
Author: Lola Marrero <110120745+lola831@users.noreply.github.com>
Date:   Mon Apr 28 14:36:36 2025 -0700

    Temporarily skip test_mesh_interiorPoint due to inconsistencies (#344)

commit 9d1e52deb0aeb479dab68cb357129fc771c283ce
Author: Daniel Fremont <dfremont@ucsc.edu>
Date:   Thu Apr 17 16:36:12 2025 -0700

    Various optimizations (#319)

    * disable redundant centering of occupiedSpace

    * apply MeshRegion transformations simultaneously

    * reduce redundant checks on occupiedSpace

    * improve backtrace for errors inside Point @properties

    * PolygonalFootprintRegion.containsObject checks convex hull before exact bounding polygon

    * speed up bounding box intersection check

    * built-in mutators directly modify objects

    * [WIP] precompute geometric data when shape/dimensions are fixed

    * precompute geometric data when shape/dimensions are fixed

    * improve requirement sorting heuristic

    * fix collision detection benchmarking

    * precompute body count of MeshVolumeRegion

    * speed up copying trimeshes when creating MeshRegions

    * avoid redundant scale computations for fixed shapes

    * reuse occupiedSpace for fixed objects

    * compute MeshRegion.mesh lazily

    * do circumradius check before computing bounding boxes; skip inradius check unless precomputed

    * skip bounding box check if meshes are lazy

    * fix transforms of precomputed Shape data; cache MultiPoint of vertices

    * minor tweaks & extra tests to cover new lines

    * minor fix; more tests

    * fix reproducibility issue

    * robustify MeshRegion boundingPolygon test

commit 4ecc20b09ce4efd200ff4d88beb16351ffd0b9bd
Author: Shreedhar Govil <shreedhargovil@gmail.com>
Date:   Thu Apr 17 00:39:00 2025 +0200

    Improve Carla Autopilot configuration for Scenic (#332)

    * add speed to Scenic Autopilot

    * add autopilot changes

    * chore: apply Black formatting to Carla autopilot changes

    ---------

    Co-authored-by: Shreedhar Govil <shreedhar.govil@dfki.de>
    Co-authored-by: lola <keepthechange2021@gmail.com>

ethan10mak committed:
Cleaned up example files

ethan10mak committed:
Elements are no longer ignored when in between other elements

ethan10mak committed:
Update xodr_parser.py

ethan10mak committed:
Updated carla\simulator.py

ethan10mak committed:
Updated carla simulator code

ethan10mak committed:
Formatted files

ethan10mak committed:
Update roads.py

ethan10mak committed:
Update roads.py

ethan10mak committed:
Updated Carla simulator

ethan10mak committed:
Update xodr_parser.py

ethan10mak committed:
Update xodr_parser.py

ethan10mak committed:
Update test_driving.py

ethan10mak committed:
Cleaned comments

ethan10mak committed:
Cleaned files

ethan10mak committed:
Cleaned files

ethan10mak committed:
Updated requirements.py

ethan10mak committed:
Update requirements.py

ethan10mak committed:
Updated requirements.py

Union Region doesn't have containsObject implemented, which causes some errors on tests, so I return True on these tests

ethan10mak committed:
Modified 2d tests

ethan10mak committed:
Cleaned file

ethan10mak committed:
Updated tests

ethan10mak committed:
Formatted file

ethan10mak committed:
Revert "Squashed commit of the following:"

This reverts commit 024e6845518b85c7a424af2f4e165d1bc26aaedb.

ethan10mak committed:
Merge branch 'main' into 3d-driving

ethan10mak committed:
Updated Test Cases, Network Element, and Network

ethan10mak committed:
Formatted Files

ethan10mak committed:
Reverted test files and updated roads with asserts

ethan10mak committed:
Updated xodr_parser.py with orientation fix to curbs

ethan10mak committed:
Updated normal vector calculation in xodr_parser

ethan10mak committed:
Updated asserts

ethan10mak committed:
Formatted roads.py

ethan10mak committed:
Reverted UnionRegion containsObject

ethan10mak committed:
Formatted files

ethan10mak committed:
Added union for MeshSurfaceRegion

ethan10mak committed:
Updated Assertions for roads.py

ethan10mak committed:
Added assertions to roads.py

ethan10mak committed:
Updated asserts and changed conftest.py to use fixture module

ethan10mak committed:
Created boundingPolygon for PolygonalRegion

There are errors in some of the tests regarding calling boundingPolygon if the region is empty.

ethan10mak committed:
Updated Assertions for roads.py

ethan10mak committed:
Added intersects function to PathRegion

ethan10mak committed:
Added fix to NetworkElement intersects and added fix to test_show2D

ethan10mak committed:
Reverted conftest from session to module

ethan10mak committed:
Updated Tests

ethan10mak committed:
Updated 3D tests, MeshSurfaceRegion, and asserts in roads.py

ethan10mak committed:
Formatted Files

ethan10mak committed:
Updated findElementWithin

ethan10mak committed:
Fixed issue with not returning the correct maneuvers for intersections

ethan10mak committed:
Fixed formatting issue

ethan10mak committed:
Updated containRegionInner for MeshSurfaceRegions

containsRegionInner() now uses the bounding polygon of the mesh surface regions, and checks if self contains reg

ethan10mak committed:
Updated default tolerance

I increased the base tolerance as some maps seem to need a slight increase in tolerance in 3D mode to pass assertions with regions

ethan10mak committed:
Updated tolerance for test_element_tolerance

ethan10mak committed:
Created new tests specific to 3D elements

ethan10mak committed:
Formatted test_driving.py

ethan10mak committed:
Update roads.py

ethan10mak committed:
Updated tests

ethan10mak committed:
Formatted test_network.py

dfremont commented on tests/domains/driving/test_driving.py:
```diff
@@ -96,6 +107,8 @@ def test_elements_at(cached_maps):
             param intersection = network.intersectionAt(spot)
     """,
+        mode2D=use2DMap,
+        params={"use2DMap": use2DMap},
```
Here and in the other tests, you shouldn't need to specify `use2DMap` since it will be turned on automatically in 2D mode and otherwise turned off.

ethan10mak committed:
Update src/scenic/core/regions.py

Co-authored-by: Daniel Fremont <dfremont@ucsc.edu>

ethan10mak committed:
Removed unnecessary parameters from tests

ethan10mak committed:
Merge branch '3d-driving' of https://github.com/BerkeleyLearnVerify/Scenic into 3d-driving

ethan10mak committed:
Updated test_driving.py

Made mode2D=False by default

ethan10mak committed:
Update model.scenic

ethan10mak committed:
Update test_basic.py

ethan10mak committed:
Update test_basic.py

ethan10mak committed:
Formatted files

ethan10mak committed:
Updated carla tests with 3D tests

(WARNING): The CARLA simulator isn't working at the time, so I have not been able to verify this code runs properly.

ethan10mak committed:
Formatted files

ethan10mak committed:
Update 3D driving CARLA tests

ethan10mak committed:
Changed test_brake back to Town01

lola831 committed:
Merge main into 3d-driving-update

lola831 committed:
Update tests for 2D/3D map options caching

lola831 committed:
carla: apply vehicle/walker spawn Z offset regardless of snapToGround

lola831 committed:
bump .snet format version

lola831 committed:
carla: use use2DMap for snapToGroundDefault

lola831 committed:
Fix NetworkElement post-init ordering so default orientation is set before Region initialization

lola831 committed:
Fix default vehicle/pedestrian placement in 3D mode

lola831 committed:
Add default pedestrian height in driving domain

lola831 committed:
carla: use scenicToCarlaTransform for spawning and placement actions (CarlaActor.setPosition and SetTransformAction)

lola831 committed:
Revert NetworkElement __attrs_post_init__ ordering change. Update driving model centerOnSurface to derive surface orientation from element at contact point

lola831 committed:
Deprecate scenicToCarlaLocation;  keep wrapper for backward compatibility

lola831 committed:
Adjust default pedestrian height

lola831 committed:
fix CARLA setPosition docstring

lola831 committed:
Clean up driving test file

lola831 committed:
stabilize CARLA throttle test

lola831 committed:
Use separate basic scenario for 3D CARLA tests

lola831 committed:
Merge main into 3d-driving and resolve conflicts

lola831 committed:
Fix 3D test_elements_at by preserving ego.position.z

lola831 committed:
Fix flaky CARLA basic_3d tests

lola831 committed:
Add specialized sampler for mesh surface-volume intersections

Handle MeshSurfaceRegion/MeshVolumeRegion intersections through a custom sampler instead of the generic intersection fallback. The sampler filters surface triangles using bounding-box overlap with the volume, then samples from the reduced surface and checks points against the volume exactly.

This improves 3D sampling enough to go back to using the same basic.scenic CARLA test scenario instead of needing a separate 3D-specific case.

Also add core region tests covering sampler wiring and sampling behavior for surface-volume intersections.

lola831 committed:
Merge main into 3d-driving-update

lola831 committed:
Update CARLA test_car_clipping_3D test
Update test_car_clipping_3D to generate a scene instead of only compiling, and allow either InvalidScenarioError or RejectionException. With dynamic CARLA blueprints, default car dimensions depend on the sampled blueprint, so overlapping cars will not be rejected during the early static validation pass and instead will fail during scene generation.

lola831 committed:
Simplify surface-volume sampler and increase CARLA test max iterations

lola831 committed:
Use tolerance-aware batched check in surface-volume sampler

lola831 commented:
The test suite was passing except for the two CARLA tests in `tests/simulators/carla/test_basic.py` that use `basic.scenic` (`test_basic[False]` and `test_car_created[False]`).

The new surface-volume sampler, `sampleSurfaceInVolume`, helped, but those 3D cases can still occasionally require more iterations to generate a scene, so I increased the iteration limit for those tests from 1000 to 2000.

dfremont commented on src/scenic/core/regions.py:
```diff
@@ -2767,11 +2881,22 @@ def __init__(
             numpy.linalg.norm(b - a, axis=1), (len(a), 1)
         )
 
+    def intersects(self, other, triedReversed=False):
```
The implementation here is not correct: it ignores the Z coordinates. Using Shapely isn't a bad idea (it's probably faster than trying to compute intersections ourselves), but you'd have to check that at least one of the 2D intersections it finds is actually an intersection in 3D. For intersection points, that's easy; for segments, I think it's enough to check each of the endpoints.

dfremont commented on src/scenic/domains/driving/model.scenic:
```diff
@@ -283,13 +307,14 @@ class Vehicle(DrivingObject):
-    regionContainedIn: roadOrShoulder
-    position: new Point on road
+    regionContainedIn: roadOrShoulder.boundingPolygon # Change region to bounding polygon region
+    position: centerOnSurface(road, self.baseOffset, self.contactTolerance)
```
I think I see why we need this now, but with a mechanism for default specifiers, this would be equivalent to simply `on road`, I think.

dfremont commented on src/scenic/domains/driving/roads.py:
```diff
@@ -136,8 +146,18 @@ def guessTypeFromLanes(
         # Identify turns based on relative heading of start and end of connecting lane
-        startDir = connecting.centerline[1] - connecting.centerline[0]
-        endDir = connecting.centerline[-1] - connecting.centerline[-2]
+        startDir = None
+        endDir = None
+        if isinstance(connecting.centerline, PathRegion):
+            startDir = (
+                connecting.centerline.vertices[0] - connecting.centerline.vertices[1]
+            )
+            endDir = (
+                connecting.centerline.vertices[-2] - connecting.centerline.vertices[-1]
+            )
+        else:
+            startDir = connecting.centerline[1] - connecting.centerline[0]
+            endDir = connecting.centerline[-1] - connecting.centerline[-2]
```
I think we can revert this change since `PathRegion.__getitem__` was added above. (Also the indices look backwards for the PathRegion case -- better double-check this!)

dfremont commented on src/scenic/domains/driving/roads.py:
```diff
@@ -246,10 +269,18 @@ def __attrs_post_init__(self):
-            polygon=self.polygon, orientation=self.orientation, name=self.name
-        )
+        if isinstance(self.region, MeshSurfaceRegion):
+            self.region.__init__(
```
Shouldn't call `__init__` again: the region is already initialized.
